### PR TITLE
Cf16/seed

### PR DESCRIPTION
**Pull request by milensski**
# Yugioh Card Data Seeding

## Description
This pull request introduces a management command to seed the database with YUGIOH card data retrieved from an external API. The purpose of this feature is to populate the database with YUGIOH card information, including card details, sets, and rarities.

## Changes
- Added a new management command, `seed`, to fetch and store YUGIOH card data from an external API.
- Implemented support for three types of data seeding: `onecard`, `staple`, and `all`.
- Fetched card data from the YUGIOH API and processed it to store in the database.
- Created or updated YugiohCard, YugiohCardRarity, YugiohCardSet, and YugiohCardInSet records.
- Changed YugiohCardSet fields
- Ensured that data is correctly associated with the "Yu-Gi-Oh!" game.
- Added logging to track the progress and any errors during the data seeding process.

## Usage
To use this feature, run the following command:
```python manage.py seed <command>```

Where `<command>` can be one of the following: `onecard`, `staple`, or `all`.

## Testing
This feature has been thoroughly tested, including API request handling, data processing, and database interaction. 
Log file is automatically created `./yugioh_seeding.log` , please let me know if its need more concrete messages.

> [!WARNING]  
> Please use the option for importing all cards with caution. 

> [!NOTE]
> The total number of cards is 12,852 and each card is in 1 or more sets. some cards are taking place in more than 59 sets.

## Reviewers
Please review this pull request and provide feedback on the code quality, documentation, and any potential improvements.

## Additional Information

> [!IMPORTANT] 
> Changes to the model YugiohCardSet is made, since the fields cannot be unique. 

**There are 2 cases:**
One case: the "set_name" is the same but the "set_code" is different
 
image

Second case: the "set_code" is the same but the "set_name" is different
 image

- The external API used for card data retrieval is provided by `https://db.ygoprodeck.com/api/v7`.
- Logging is implemented to help monitor the progress and diagnose issues.
- The data seeding process is designed to be robust, handling key errors and unexpected data gracefully.

Please let me know if you need any further information or if there are any specific aspects of the code that you would like to discuss.

### CODE REVIEW

File: backend/yugioh/management/commands/seed.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+import requests
+from card.models import Card
+from django.core.management.base import BaseCommand
+from game.models import Game
+from psycopg2 import DataError
+from yugioh.models import YugiohCard, YugiohCardRarity, YugiohCardSet, YugiohCardInSet
+
+log_filename = "yugioh_seeding.log"
+logging.basicConfig(filename=log_filename, level=logging.INFO,
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:11]: No need to write the logs to a file. I think it's better to log them to the output, so they can be immediately seen, also captured by the docker logs in the near future.

File: backend/yugioh/management/commands/seed.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+        api_url = self.API_URLS.get(options['command'])
+
+        if not api_url:
+            self.stdout.write(self.style.ERROR(f'Invalid command: {options["command"]}. Please check help options.'))
+            logging.error(f'Invalid command: {options["command"]}. Please check help options.')
+            return
+
+        # self.stdout.write('Calling the external API')
+        logging.info('Calling the external API')
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:65]: Be more specific of the URL that is being called.

File: backend/yugioh/management/commands/seed.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+        response = requests.get(api_url)
+
+        if response.status_code == 200:
+            #             self.stdout.write('API response received successfully')
+            logging.info('API response received successfully')
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:71]: Be more specific. What response did you receive, from where?

**Comment by julkascript** [comment]
- [ ] In django admin `Yugioh card in sets` put the card name as a first column

**Comment by julkascript** [comment]
On a clean database if I try to execute `python manage.py seed onecard` it starts seeding, it gets to somewhere, it populates objects, but at some point raises exceptions:

```
Traceback (most recent call last):
  File "/home/julka/cardflow/backend/manage.py", line 22, in <module>
    main()
  File "/home/julka/cardflow/backend/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
  File "/home/julka/cardflow/backend/yugioh/management/commands/seed.py", line 225, in handle
    rarity_object = YugiohCardRarity.objects.create(
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/query.py", line 658, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/base.py", line 814, in save
    self.save_base(
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/base.py", line 877, in save_base
    updated = self._save_table(
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/base.py", line 1020, in _save_table
    results = self._do_insert(
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/base.py", line 1061, in _do_insert
    return manager._insert(
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/query.py", line 1805, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1822, in execute_sql
    cursor.execute(sql, params)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 102, in execute
    return super().execute(sql, params)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 84, in _execute
    with self.db.wrap_database_errors:
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/julka/cardflow/backend/venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 89, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.IntegrityError: duplicate key value violates unique constraint "yugioh_yugiohcardrarity_rarity_code_key"
DETAIL:  Key (rarity_code)=() already exists.
```

File: backend/yugioh/management/commands/seed.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            frame_type = item['frameType']
+            desc = item['desc']
+
+            logging.info(f'Reading ** {card_name} ** Details')
+
+            if len(desc) >= 1000:
+                desc = item['desc'][:990] + '...'
+
+            try:
+                atk = item['atk']
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:99]: Just skip if there is not `atk` or `def`, no need to print info.

File: backend/yugioh/management/commands/seed.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            logging.info('API response received successfully')
+            data = response.json()
+        else:
+            #         self.stdout.write(self.style.ERROR(f'API request failed with status code {response.status_code}'))
+            logging.error(f'API request failed with status code {response.status_code}')
+            return
+
+        if not Game.objects.filter(game_name="Yu-Gi-Oh!").exists():
+            #             self.stdout.write('Game does not exist... Creating')
+            logging.info('Game does not exist... Creating')
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:80]: Be more specific, which game object are you creating?

File: backend/yugioh/management/commands/seed.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                        yugioh_card=yugioh_card_object
+                    ).first()
+
+                    if not existing_card:
+                        YugiohCardInSet.objects.create(rarity=rarity_object, set=card_set_object,
+                                                       yugioh_card=yugioh_card_object)
+                    else:
+                        # self.stdout.write(self.style.ERROR(f'Card {yugioh_card_object} '
+                        #                                    f'{card_set_object} {rarity_object}... Skipping'))
+                        logging.info(f'Card {yugioh_card_object} with '
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:232]: No need to be so specific with the logging.

File: backend/yugioh/management/commands/seed.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                                     f'{card_set_object} {rarity_object}... Skipping')
+            except KeyError:
+                logging.info(f'No sets for {card_name}')
+
+        # self.stdout.write(self.style.SUCCESS(f'DONE!'))
+        logging.info('DONE!')
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:238]: The end print is usually good to provide the user with a super compact summary of what happened, e.g., how many objects were created, etc.

**Comment by milensski** [comment]
Please review the code now, i have added some logic if there is no rarity codes to be generated also there was single record where the name for Common was just 'c' and i have handled it again with if statement.
After the last commit `all` command has imported and processed everything
image

They were processed 12850 cards while in the website database it says 12852 cards, I hope this is not a big problem 
>[!NOTE]
>The import of all cards took me around 15 minutes

File: backend/yugioh/management/commands/seed.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                                                           f'{card_set_object} {rarity_object}... Skipping'))
+                        logging.info(f'Card {yugioh_card_object} with '
+                                     f'{card_set_object} {rarity_object}... Skipping')
+            except KeyError:
+                logging.info(f'No sets for {card_name}')
+
+        logging.info("--------------")
+        logging.info(f"Processed cards -> {processed_cards}")
+        logging.info(f"    - Imported cards -> {imported_card}")
+        logging.info(f"Processed sets -> {processed_sets}")
```
Comment (julkascript) [inline backend/yugioh/management/commands/seed.py:236]: The `processed_sets` value actually represents the `CardInSet` objects. The processed sets are ~980.

Comment (milensski) [reply backend/yugioh/management/commands/seed.py]: it is fixed now

**Review by julkascript — APPROVED** [review]

**Pull request merged**